### PR TITLE
iOS: rebuild terminal keyboard pinning on one geometry authority

### PR DESCRIPTION
**Pull request by azooz2003-bit**
Reimplements the iOS terminal keyboard pinning from scratch so the composer bar, shortcuts bar, and terminal bottom track the keyboard with no bad frames.

## Why a rebuild

Five open PRs (#9663, #9770, #9827, #10147, #10158) patched the same symptom set: one-frame seams between the dock and the terminal on interrupted reversals, blank terminal exposed during keyboard rise, a content snap after keyboard dismissal, and bars wedged under a raised keyboard after workspace switches. Each traced back to two structural choices on main: `UIKeyboardLayoutGuide` as the dock authority (misses transitions while the view is detached, broken on iOS 27, second animation clock), and keyboard motion split across two mechanisms (a constraint for the dock, a `CGAffineTransform` for the renderer) that required presentation-layer rebasing to stay aligned.

## What changed

- **One geometry authority.** Keyboard frame notifications drive the dock on every OS version. The `UIKeyboardLayoutGuide` path, the iOS 27 fork (`KeyboardDockGeometrySource`), and `CMUX_UITEST_FORCE_IOS27_KEYBOARD_DOCK` are deleted.
- **One moving constraint system.** A keyboard leg changes two constraint constants (dock bottom, render-wrapper bottom) and animates a single `layoutIfNeeded()` on the keyboard's curve. The clip boundary between terminal and bars is a constraint to the dock top, so bars and terminal boundary derive from the same layout pass and cannot land on different timelines. `.beginFromCurrentState` retargets reversals from live presentation frames; `rebaseKeyboardPresentationFromLiveFrames` and the dock-animation stripping hooks are deleted.
- **Animated endpoint = settled layout.** The wrapper animates the render bottom to the exact edge the first post-transition layout pass computes (same snapshot + cursor-aware `renderPinnedBottomEdge` math), so the completion fold is a visual no-op. Previously the renderer was translated unconditionally to the dock edge and the next layout pass snapped it to the cursor-aware pin.
- **Detached-transition recovery.** New `MobileKeyboardFrameTracker` records keyboard end frames process-wide (armed at launch); a host attaching to a window settles its dock and the renderer model from that record. This also clears a stale `keyboardPresentationTransitionActive` left by a leg interrupted by detach, which could previously freeze grid negotiation until the next keyboard event.
- **Tests.** `TerminalKeyboardDockEndpoints` carries the endpoint arithmetic with 5 unit tests. `testIOS27KeyboardDockWorkaroundPinsComposerToKeyboard` becomes `testNotificationKeyboardDockPinsComposerToKeyboard`, the universal contract test; the dock probe keeps its `keyboardDockSource` key (always `notification`). Existing reversal XCUITests (`testTerminalDockStaysUnifiedAcrossRapidKeyboardReversals`, `testTerminalDockStaysPinnedForInPlaceKeyboardControlReversals`) exercise the new path unchanged.

Localization audit: no user-facing strings added or changed (all internal layout code).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
<a href="https://app.blacksmith.sh/manaflow-ai/codesmith/cmux/pr/10518"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with [code]smith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1789886889&installation_model_id=21920&pr_number=10518&repository=manaflow-ai%2Fcmux&return_to=https%3A%2F%2Fgithub.com%2Fmanaflow-ai%2Fcmux%2Fpull%2F10518&signature=4167988f0e32c4bb7e09c69e8a9628b203702f46a1c74bb054460929eb88ad45"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with [code]smith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith-bot</code> with what you need. Autofix is disabled.</sup>

---
## Summary by cubic
Rebuilds iOS terminal keyboard pinning around a single notification‑driven geometry authority and removes `UIKeyboardLayoutGuide`. Ships the legacy notification+transform path as the default on all OS versions; iOS ≤26 can revert to the rebuilt single‑constraint path for correctness and stability.

- Keyboard geometry and animation: notifications drive two constraint constants (dock bottom, render-wrapper bottom) in one keyboard-curve `layoutIfNeeded()`; animated endpoint equals the first settled layout; `.beginFromCurrentState` handles reversals; `keyboardDidChangeFrame` retargets only on disagreement with a short curve; layout self-heals the dock seat from an app‑lifetime `MobileKeyboardFrameTracker` outside active legs; detach strips keyboard-motion animations to avoid resuming stale legs.
- Legacy path (default everywhere; always on iOS 27+): will‑frame only; dock by constraint plus wrapper transform; interrupted legs rebase from presentation layers; fold pins the render at the dock; both paths recover detach from the tracker; probe reports `keyboardDockSource=legacyNotification` (rebuilt path reports `notification`).
- Tracker and injection: adds `MobileKeyboardFrameTracker` in `CmuxMobileSupport`; injected via `\.mobileKeyboardFrameTracker` from the app composition root so hosts recover transitions missed while detached; previews/harnesses use a coordinator fallback.
- Path selection and controls: `TerminalKeyboardDockPathSelection` selects the path once at host mount. Remote kill switch `ClientConfigFlag.iosKeyboardDockRebuildRevert` (cached in `MobileFeatureFlags`, injected via `\.keyboardDockRebuildRevertEnabled`) routes iOS ≤26 to the rebuild; iOS 27+ never routes to it. DEBUG overrides: `CMUX_UITEST_FORCE_LEGACY_KEYBOARD_DOCK`, `CMUX_UITEST_FORCE_REBUILD_KEYBOARD_DOCK`, and Settings > Developer “Rebuilt Keyboard Pinning” (persists `-cmux.mobile.debug.forceRebuildKeyboardDock.v1`; reopen the workspace to apply; release builds ignore).
- Tests and metrics: extracts endpoint math to `TerminalKeyboardDockEndpoints` with unit tests in `CmuxMobileTerminalKit`; adds unit tests for path selection; XCUITests assert the clip‑to‑dock seam stays zero and that the dock sits within the keyboard’s accessory chrome band; render‑attachment checks move to echo‑settled points; probe adds renderer/dock gap metrics. Localization adds settings strings only.

<sup>Written for commit 5b504760f78d64af545b159f3284ac26e9b24122. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/manaflow-ai/cmux/pull/10518?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Added reliable keyboard frame tracking and visibility detection.
  * Improved terminal docking and rendering during keyboard transitions, including detached views and oversized keyboards.
  * Added consistent keyboard geometry calculations across supported iOS versions.

* **Bug Fixes**
  * Improved keyboard animations, dismissal handling, and bottom-dock positioning.
  * Removed reliance on OS-specific keyboard behavior and test overrides.

* **Tests**
  * Added coverage for keyboard movement, dismissal, partial overlap, and boundary conditions.

### CODE REVIEW

File: Packages/iOS/CmuxMobileSupport/Sources/CmuxMobileSupport/MobileKeyboardFrameTracker.swift [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+    /// The keyboard's most recent end frame in screen coordinates, or `nil`
+    /// before the first keyboard notification of the process.
+    public private(set) var lastEndFrame: CGRect?
+
+    private nonisolated(unsafe) var tokens: [NSObjectProtocol] = []
+    private nonisolated let notificationCenter: NotificationCenter
+
+    /// Creates a tracker subscribed to the keyboard frame notifications.
+    public init(notificationCenter: NotificationCenter = .default) {
```
[the remark this answers is not in the document]
Comment (azooz2003-bit) [reply Packages/iOS/CmuxMobileSupport/Sources/CmuxMobileSupport/MobileKeyboardFrameTracker.swift:34]: The recorded frame is screen-space exactly as UIKit broadcasts keyboard notifications to every scene in the process, and each consumer converts it through its OWN window and intersects it with its own bounds (MobileKeyboardReservation), so a multi-scene read yields the overlap the keyboard actually has over that scene's window — a scene the keyboard does not cover resolves to zero reservation. Per-scene visibility drift observed while a host was detached self-corrects on the next keyboard notification in that scene, the same convergence UIKit's own process-wide broadcast relies on. Documented on the tracker in 6c2d7d81c0.

**Comment by azooz2003-bit** [comment]
Verification round complete on head 6c2d7d81c0 (app-behavior code e9f26c40f8):

- All three keyboard XCUITests pass on CI (job `ios-simulator (iphone)`): [rapid reversal](https://github.com/manaflow-ai/cmux/actions/runs/32527267596), [notification dock](https://github.com/manaflow-ai/cmux/actions/runs/32527272956), [in-place reversal](https://github.com/manaflow-ai/cmux/actions/runs/32527278469). The failing `package-conventions-lint` job flags six pre-existing files absent from this PR's diff (main is red on it); `ios-tests` fails only as its aggregator.
- Independent frame-split verification on an isolated simulator (2,230 frames over five scenarios: show, hide, rapid A→B→A reversals, workspace re-entry, content-snap): zero seam pixels between the terminal boundary and the bars on every frame, ≤1px bars↔keyboard, exact 25px steady inset, and the in-app per-display-frame samplers read dockMaxInternalPresentationGap=0.000 / terminalDockMaxPresentationGap=0.000 across every transition. CPU peaks at 37% per transition, no sustained main-thread load.
- Follow-up commits during verification: layout self-heal re-derives the dock seat from the tracker when the host's bounds change after a keyboard notification (fixes a 17pt stale-bounds seat CI caught); a disagreement-gated `keyboardDidChangeFrame` observer corrects keyboard re-seats; the per-frame seam metric now measures the terminal CLIP edge (the structural contract) with the renderer distance kept as informational keys; the UITest keyboard oracle asserts the dock inside the keyboard's accessory-chrome band instead of equality with the bare key plane.

Evidence bundle: `out/kbpp-verify/evidence/` in the dogfood worktree (videos, frame CSVs, contact sheets, probe dumps, profiling).

**Comment by azooz2003-bit** [comment]
iOS 27 scope update, verified on-device-sim: the rebuild is now gated to iOS 26 and earlier. iOS 27+ keeps the pre-rebuild notification+transform dock path (`usesLegacyKeyboardDockPath`, probe reports `keyboardDockSource=legacyNotification`). Verified on an iOS 27.0 simulator with the real signed-in app at 225f37ee68: keyboard toggle up and down both work, composer bottom sits exactly at the notification keyboard top (419 = 720-301), terminal render bottom exactly at dock top, `terminalDockPresentationGap=0.000` and `dockInternalPresentationGap=0.000` in both directions. CI on this head: notification dock, forced-legacy dock, and 10-cycle rapid-reversal tests all green (runs 32695190260 / 32695194057 / 32695198314).
